src/Lecture2.hs

```diff
@@ -152,12 +183,124 @@ You're free to define any helper functions.
 
 -- some help in the beginning ;)
 data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
-    , knightEndurance :: Int
-    }
+    { knightHealth    :: Health
+    , knightAttack    :: Attack
+    , knightEndurance :: Endurance
+    } deriving (Eq)
 
-dragonFight = error "TODO"
+data Dragon = Dragon
+    { dragonHealth    :: Health
+    , dragonFirePower :: FirePower
+    } deriving (Eq)
+
+data Chest a = Chest
+    { gold      :: Gold
+    , threasure :: Maybe a
+    } deriving (Show, Eq)
+
+data Reward = Reward Gold (Maybe Treasure) XP deriving (Show, Eq)
+
+data Treasure = Gems | Jewels deriving (Show, Eq)
+
+data EvilDragon = Green Dragon Gold
+                | Black Dragon (Chest Treasure)
+                | Red   Dragon (Chest Treasure) deriving (Eq)
+
+data FightResult = KnightWin Reward | KnightDie | KnightRunAway deriving (Show, Eq)
+
+newtype Gold = Gold Int deriving (Show, Eq)
+
+newtype Health = Health Int deriving (Show, Eq)
+
+newtype Attack = Attack Int deriving (Show, Eq)
+
+newtype Endurance = Endurance Int deriving (Show, Eq)
+
+newtype FirePower = FirePower Int deriving (Show, Eq)
+
+newtype XP = XP Int deriving (Show, Eq)
+
+newtype Error = Error String deriving (Show, Eq)
+
+dragonFight :: Knight -> EvilDragon -> Either Error FightResult
+dragonFight knight evilDragon = evaluateResult (snd fightResult) (fst fightResult)
+    where go :: Int -> Knight -> EvilDragon -> (Knight, EvilDragon)
+          go strikeNumber knight' dragon
+            | isFightFinished evilDragon knight' = (knight', dragon)
+            | strikeNumber > maxStrikesInRound   = go 1 (damageKnight dragon knight') dragon
+            | otherwise                          = go (strikeNumber + 1) (reduceEndurance knight') (damageDragon knight' dragon)
+          fightResult :: (Knight, EvilDragon)
+          fightResult = go 1 knight evilDragon
+
+maxStrikesInRound :: Int
+maxStrikesInRound = 10
+
+evaluateResult :: EvilDragon -> Knight -> Either Error FightResult
+evaluateResult evilDragon knight
+  | isDead (dragonHealth dragon) = Right $ KnightWin $ createReward evilDragon
+  | isEnduranceEmpty (knightEndurance knight) = Right KnightRunAway
+  | isDead (knightHealth knight) = Right KnightDie
+  | otherwise = Left $ Error "something went wrong :("
+  where dragon = extractDragon evilDragon
+
+isFightFinished :: EvilDragon -> Knight -> Bool
+isFightFinished dragon knight = isDead (knightHealth knight)
+                             || isDead (dragonHealth (extractDragon dragon))
+                             || isEnduranceEmpty (knightEndurance knight)
```

Comment:
These two functions contain duplicate code. In your code, you check for the condition to stop the fight but then you check the same conditions again to see the final result. This is a case of a thing known as "boolean blindness". You've gained lots of knowledge about your values but you reduced them to a single boolean value. On top of that, you have impossible `otherwise` case in `evaluateResult` to cover situations that should never happen. 

Instead, it's better to return the result immediately: if one of the conditions is true, you return the result. Otherwise, you call the function recursively 🔃

Reply:
Thanks for the feedback, I will try a refactor before a merge 👍

Reply:
I finished the refactor, totally worth it... removed the nonsense Error and a bunch of functions. Thanks for the tip!